extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -44,54 +45,96 @@ export class AmpViewerIntegration {
 
     /** @private {?string|undefined} */
     this.unconfirmedViewerOrigin_ = null;
+
+    /** @private {boolean} */
+    this.isWebView_ = false;
   }
 
   /**
    * Initiate the handshake. If handshake confirmed, start listening for
    * messages. The service is disabled if the viewerorigin parameter is
    * absent.
-   * @return {?Promise}
+   * @return {!Promise<undefined>}
    */
   init() {
     dev().fine(TAG, 'handshake init()');
     const viewer = viewerForDoc(this.win.document);
+    this.isWebView_ = viewer.getParam('webview') == 1;
     this.unconfirmedViewerOrigin_ = viewer.getParam('origin');
-    if (!this.unconfirmedViewerOrigin_) {
-      dev().fine(TAG, 'Viewer origin not specified.');
-      return null;
+
+    if (this.isWebView_) {
+      return this.webviewPreHandshakePromise_(null, dev().assertString(''))
+      .then(receivedPort => {
+        return this.openChannelAndStart_(viewer, receivedPort);
+      });
     }
 
-    dev().fine(TAG, 'listening for messages', this.unconfirmedViewerOrigin_);
-    const messaging = new Messaging(
-      this.win, this.win.parent, this.unconfirmedViewerOrigin_);
+    const port = new WindowPortEmulator(
+      this.win, dev().assertString(this.unconfirmedViewerOrigin_));
+    return this.openChannelAndStart_(viewer, port);
+  }
 
+  /**
+   * @return {!Promise}
+   * @param {?Window} source
+   * @param {string} origin
+   * @private
+   */
+  webviewPreHandshakePromise_(source, origin) {
+    dev().fine(TAG, origin); //delete this line before merging.
+    return new Promise(resolve => {
+      listen(this.win, 'message', e => {
+        dev().fine(TAG, 'AMPDOC got a pre-handshake message:', e.type, e.data);
+        // Viewer says: "I'm ready for you"
+        if (
+            // e.origin === origin && //uncomment before submit
+            // e.source === source && //undomment before submit
+            e.data.app == APP &&
+            e.data.name == 'handshake-poll') {
+          if (!e.ports || !e.ports.length) {
+            throw new Error(
+              'Did not receive communication port from the Viewer!');
+          }
+          resolve(e.ports[0]);
```

Comment:
We also normally need to unlisten since we are no longer interested in these messages.